### PR TITLE
Introduce a configuration manager

### LINKED ISSUE #464: Plugin settings aren't write in plugin folder
By design, `plugins/<plugin>/config.php` overrides `config.php` for this specific plugin.But when you save settings in the plugin admin page, the plugin config file is not set (it only has effect on global `config.php`).

### PR DESCRIPTION
**Pull request by ArthurHoaro**
> Warning: this PR is huge, even if I try to split commits, since it touches the whole codebase.

This introduces a configuration manager which let us read and write settings anywhere properly:

```php
$conf = ConfigManager::getInstance();
$foo = $conf->get('settings.foo', 'default');
$foo = 'bar';
$conf->set('settings.foo', $foo);
$conf->write(isLoggedIn());
```
It will allow us to get rid of the `$GLOBALS` configuration system, which is a very bad practice. We'll switch to JSON according to #489. 

This PR doesn't contain JSON conf yet, but proxy the PHP conf file behind `ConfigManagement`, for transition.

Done: 

  * [x] JSON settings
  * [x] Use the updater to create an initial JSON settings file from `config.php`
  * [x] Set the timezone
  * [x] Update plugins with configuration
  * [x] More robust unit tests
  * [x] Rename and group configuration keys.
  * [x] Documentation
  * [x] Remove remaining config initialization from `index.php`.

closes #471 closes #464

### CODE REVIEW

**Comment by virtualtam** [comment]
Cool! Does this PR obsolete #464 and #471 ?

**Comment by ArthurHoaro** [comment]
Yes it does. We'll keep everything in a single file. I've added plugin update in the todo list.

**Comment by virtualtam** [comment]
An other point to check: prevent the JSON configuration file from being accessed from a browser. This will require users to update their server / virtualhost configuration, or to store the JSON data in a serialized PHP script.

**Comment by ArthurHoaro** [comment]
The access should be denied for the whole `data/` directory (that's what the `.htaccess` file is doing). But we still can wrap the JSON data in a PHP file to prevent badly configured servers to expose their credentials.

**Comment by virtualtam** [comment]
htaccess is Apache-specific; we need to handle other setups as well (nginx, lighthttpd) and provide a sane default configuration in the docs (the lighthttpd section is empty btw, I'll see if I can fix this)

**Comment by nicolasdanelon** [comment]
And what about store de json between php tags?

**Comment by ArthurHoaro** [comment]
@virtualtam There is no htaccess equivalent for nginx, but we have a pretty good documentation on it (thanks to you I believe). I never used another webserver.

@nicolasdanelon Yep that's what we meant.

**Comment by nicolasdanelon** [comment]
Oh great! Good luck then ^^

**Comment by ArthurHoaro** [comment]
It's starting to look good. Please review and test it, since I might have missed a few things in all of this.

I've renamed configuration keys to be a bit more coherent. The documentation is [here](https://github.com/ArthurHoaro/Shaarli/blob/6a080b498891037dabcd96bcef157a2718558672/doc/Shaarli-configuration.md). It will be added in the wiki when this PR is merged.

Also note that the automatic transition PHP -> JSON code is a bit messy. It might be a good idea to drop it when we'll reach 1.0.0.

**Comment by nodiscc** [comment]
Does it migrate current user settings to the new format? Or does the user have to reconfigure/reinstall? Great work!

**Comment by ArthurHoaro** [comment]
Everything is automagic, and the PHP conf file is backed up.

**Comment by virtualtam** [comment]
Neat! I'm starting the review ;-)

File: application/config/ConfigManager.php [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+        $this->setEmpty('path.updates', 'data/updates.txt');
+        $this->setEmpty('path.log', 'data/log.txt');
+        $this->setEmpty('path.update_check', 'data/lastupdatecheck.txt');
+        $this->setEmpty('path.raintpl_tpl', 'tpl/');
+        $this->setEmpty('path.raintpl_tmp', 'tmp/');
+        $this->setEmpty('path.thumbnails_cache', 'cache');
+        $this->setEmpty('path.page_cache', 'pagecache');
+
+        $this->setEmpty('security.ban_after', 4);
+        $this->setEmpty('security.ban_after', 1800);
```
Comment (virtualtam) [inline application/config/ConfigManager.php]: ban_after is reset twice

File: application/config/ConfigManager.php [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+        }
+        $conf[$setting] = $value;
+    }
+
+    /**
+     * Set a bunch of default values allowing Shaarli to start without a config file.
+     */
+    protected function setDefaultValues()
+    {
+        $this->setEmpty('path.data_dir', 'data');
```
Comment (virtualtam) [inline application/config/ConfigManager.php]: maybe group entries semantically instead of resource type?

- resources
  - data_dir
  - datastore
  - cache
  - ...
- display
  - hide_timestamps
  - links_per_page
- feed
  - rss_permalinks
  - show_atom
- security
  - ban_file
  - ban_after
  - default_private_links
  - open_shaarli
...

this feels a bit messy though, as some entries could belong to several categories

File: application/config/ConfigPhp.php [hunk 3]
```diff
@@ -0,0 +1,132 @@
+<?php
+
+/**
+ * Class ConfigPhp (ConfigIO implementation)
+ *
+ * Handle Shaarli's legacy PHP configuration file.
```
Comment (virtualtam) [inline application/config/ConfigPhp.php]: Some 3rd party plugins / themes use `config.php` to store additional settings, e.g. @kalvn 's https://github.com/kalvn/Shaarli-Material
Could we attempt to migrate these values, or should plugin / theme developers make the transition to the new config system?

File: doc/Shaarli-configuration.md [hunk 4]
```diff
@@ -1,14 +1,17 @@
-#Shaarli configuration
+# Shaarli configuration
```
Comment (virtualtam) [inline doc/Shaarli-configuration.md]: the local doc is a mere mirror of the GH pages, it's a bit risky to include these changes in a PR, as:
- they may be overwritten
- there may be conflicts with the changes made while this PR is being reviewed

to keep track of the necessary documentation changes, it't be more convenient to open a dedicated, temporary PR to highlight the diffs

File: index.php [hunk 5]
```diff
@@ -166,10 +59,12 @@
 require_once 'application/TimeZone.php';
 require_once 'application/Url.php';
 require_once 'application/Utils.php';
-require_once 'application/Config.php';
+require_once 'application/config/ConfigManager.php';
```
Comment (virtualtam) [inline index.php]: import sorting

File: index.php [hunk 6]
```diff
@@ -345,31 +234,33 @@ function pubsubhub()
...
-    $ip = $_SERVER["REMOTE_ADDR"];
+    $ip = $_SERVER['REMOTE_ADDR'];
     // Then we use more HTTP headers to prevent session hijacking from users behind the same proxy.
     if (isset($_SERVER['HTTP_X_FORWARDED_FOR'])) { $ip=$ip.'_'.$_SERVER['HTTP_X_FORWARDED_FOR']; }
     if (isset($_SERVER['HTTP_CLIENT_IP'])) { $ip=$ip.'_'.$_SERVER['HTTP_CLIENT_IP']; }
     return $ip;
 }
 
 function fillSessionInfo() {
+    $conf = ConfigManager::getInstance();
```
Comment (virtualtam) [inline index.php]: why not instantiating a ConfigManager once in `index.php`, and injecting this instance as a paramater to all functions/methods requiring it?

this may also avoid creating a Singleton :)

File: index.php [hunk 7]
```diff
@@ -395,44 +286,64 @@ function logout() {
...
-        $gb['BANS'][$ip]=time()+$GLOBALS['config']['BAN_DURATION'];
-        logm($GLOBALS['config']['LOG_FILE'], $_SERVER['REMOTE_ADDR'], 'IP address banned from login');
+        $gb['BANS'][$ip] = time() + $conf->get('security.ban_after', 1800);
+        logm($conf->get('path.log'), $_SERVER['REMOTE_ADDR'], 'IP address banned from login');
     }
     $GLOBALS['IPBANS'] = $gb;
-    file_put_contents($GLOBALS['config']['IPBANS_FILENAME'], "<?php\n\$GLOBALS['IPBANS']=".var_export($gb,true).";\n?>");
+    file_put_contents(
+        $conf->get('path.ban_file', 'data/ipbans.php'),
+        "<?php\n\$GLOBALS['IPBANS']=".var_export($gb,true).";\n?>"
```
Comment (virtualtam) [inline index.php]: could be worth raising a new issue to store banned IPs in a JSON file

File: application/config/ConfigPhp.php (hunk 3 above)
Comment (ArthurHoaro) [inline application/config/ConfigPhp.php]: The whole `$GLOBALS['config']` and `$GLOBALS['plugins']` will be converted in JSON.

File: doc/Shaarli-configuration.md (hunk 4 above)
Comment (ArthurHoaro) [inline doc/Shaarli-configuration.md]: Alright, I'll do that.

File: index.php (hunk 6 above)
Comment (ArthurHoaro) [inline index.php]: Alright, that's probably better.

File: application/config/ConfigManager.php (hunk 2 above)
Comment (ArthurHoaro) [inline application/config/ConfigManager.php]: What about this

------------------
## Configuration

In `data/config.json.php`.

### Credentials
 
> You shouldn't edit those.

**login**: Login username.

**hash**: Generated password hash.

**salt**: Password salt.

### Security

**session_protection_disabled**: Disable session cookie hijacking protection (not recommended). 
It might be useful if your IP adress often changes.

**ban_after**: Failed login attempts before being IP banned.

**ban_duration**: IP ban duration in seconds.

**open_shaarli**: Anyone can add a new link while logged out if enabled.

### Resources

**data_dir**: Data directory.

**datastore**: Shaarli's links database file path.

**updates**: File path for the ran updates file.

**log**: Log file path.

**update_check**: Last update check file path.

**raintpl_tpl**: Templates directory.

**raintpl_tmp**: Template engine cache directory.

**thumbnails_cache**: Thumbnails cache directory.

**page_cache**: Shaarli's internal cache directory.

**ban_file**: Banned IP file path.


### General

**title**: Shaarli's instance title.

**header_link**: Link to the homepage.

**links_per_page**: Number of shaares displayed per page.

**enabled_plugins**: List of enabled plugins.

### Updates

**check_updates**: Enable or disable update check to the git repository.

**check_updates_branch**: Git branch used to check updates (e.g. `stable` or `master`).

**check_updates_interval**: Look for new version every N seconds (default: every day).

### Privacy

**default_private_links**: Check the private checkbox by default for every new link.

**hide_public_links**: All links are hidden while logged out.

**hide_timestamps**: Timestamps are hidden.

### Feed

**rss_permalinks**: Enable this to redirect RSS links to Shaarli's permalinks instead of shaared URL.

**show_atom**: Display ATOM feed button.

### Thumbnail

**enable_thumbnails**: Enable or disable thumbnail display.

**enable_localcache**: Enable or disable local cache.

### Redirector

**url**: Redirector URL, such as `anonym.to`.

**encode_url**: Enable this if the redirector needs encoded URL to work properly.

File: index.php (hunk 7 above)
Comment (ArthurHoaro) [inline index.php]: #587 created

**Comment by ArthurHoaro** [comment]
The configuration manager no longer uses the singleton pattern, and is injected everywhere. I agree, it's much cleaner. :smile: 

Note that this touches the PluginManager. The `ConfigManager` is now injected in every plugin hook, which changes their signature. E.g.

```php
function hook_wallabag_render_linklist($data, $conf)
```
Third party plugins using configuration need to be updated accordingly (ping @kalvn). The `$conf` parameter is optional if not used.

I've also change the behaviour of PluginManager. It's no longer singleton either, and the configuration manager is injected in the constructor as a reference.

I've run a bunch of tests, everything looks good.

**Comment by kalvn** [comment]
It looks nice, I'll update my plugins accordingly.

File: application/config/ConfigManager.php [hunk 8]
```diff
@@ -87,7 +65,7 @@ public function reload()
      */
     protected function initialize()
     {
-        if (! file_exists(self::$CONFIG_FILE .'.php')) {
+        if (! file_exists($this->configFile .'.php')) {
```
Comment (virtualtam) [inline application/config/ConfigManager.php]: to help readability:

```php
if (file_exists($this->configFile . '.json')) {
    $this->configIO = new ConfigJson();
} else {
    $this->configIO = new ConfigPhp();
}
```

File: tests/plugins/PluginReadityourselfTest.php [hunk 9]
```diff
@@ -4,6 +4,8 @@
  * PluginReadityourselfTest.php.php
  */
 
+// FIXME! add an init method.
```
Comment (virtualtam) [inline tests/plugins/PluginReadityourselfTest.php]: to be removed?

File: tests/plugins/PluginWallabagTest.php [hunk 10]
```diff
@@ -4,6 +4,8 @@
  * PluginWallabagTest.php.php
  */
 
+// FIXME! add an init method.
```
Comment (virtualtam) [inline tests/plugins/PluginWallabagTest.php]: to be removed?

File: application/config/ConfigManager.php (hunk 2 above)
Comment (virtualtam) [inline application/config/ConfigManager.php]: doge-much-clear

**Comment by virtualtam** [comment]
> The configuration manager no longer uses the singleton pattern, and is injected everywhere. I agree, it's much cleaner.

It _sometimes_ feels good to use a Singleton. It _always_ feels good to get rid of it :smile_cat: 

The overall PR looks very good, beside some minor comments. As it brings major changes to how Shaarli is configured, I'd suggest we tag and release a new (stable?) revision and bump the version code to `0.8.x` before merging it.

**Comment by ArthurHoaro** [comment]
> What about themes? How will I be able to get config from within templates?

Crap, you're using [custom settings](https://github.com/kalvn/Shaarli-Material/#customization). Templates aren't supposed to access out of scope data. For the time being, I can pass the configuration manager instance to every templates, even though it's a bad practice. Then, when we'll add a proper theme management system, we'll also add a theme configuration. @virtualtam opinion?

> I'd suggest we tag and release a new (stable?) revision and bump the version code to 0.8.x before merging it.

There has been 3 "real" commits since the last release. We can just rename 0.7.1 to 0.8.0 (and bump the next milestones).

File: application/config/ConfigManager.php (hunk 8 above)
Comment (ArthurHoaro) [inline application/config/ConfigManager.php]: It might be more readable, but it's not the same logic. But this is:

```php
if (file_exists($this->configFile . '.php')) {
    $this->configIO = new ConfigPhp();
} else {
    $this->configIO = new ConfigJson();
}
```

File: tests/plugins/PluginWallabagTest.php (hunk 10 above)
Comment (ArthurHoaro) [inline tests/plugins/PluginWallabagTest.php]: Well, it's not fixed, and I'd prefer to do that in a separate PR.

**Comment by virtualtam** [comment]
> rename 0.7.1 to 0.8.0 (and bump the next milestones)

+1, this way we can flag `0.8.x` as introducing breaking changes for 3rd-party plugins and theme customizations, and stick to only handling core/vanilla Shaarli settings migration from serialized PHP to JSON :)

File: tests/utils/config/configJson.json.php [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        "foo": "bar"
+    },
+    "resource": {
+        "datastore": "tests\/utils\/config\/datastore.php",
+        "data_dir": "tests\/utils\/config"
+    },
+    "plugins": {
+        "WALLABAG_VERSION": 1
+    }
+}
```
Comment (virtualtam) [inline tests/utils/config/configJson.json.php]: missing comment closing tag `*/`

Comment (ArthurHoaro) [inline tests/utils/config/configJson.json.php]: Isn't it useless?

**Comment by ArthurHoaro** [comment]
@kalvn I've added the configuration manager to templates variables.

If you have in `config.php`
```php
$GLOBALS['config']['foo'] = 'bar';
```
You can display it in the templates with:
```
{$conf->get('config.foo')}
```
When the updater will run, you'll end up with
```json
"config": {
     "foo": "bar"
}
```

Note that this won't work with settings outside of `config` or `plugins` array in `config.php`.

File: tests/utils/config/configJson.json.php (hunk 11 above)
Comment (virtualtam) [inline tests/utils/config/configJson.json.php]: It is, yet generating valid files would do no harm ;-)

Comment (ArthurHoaro) [inline tests/utils/config/configJson.json.php]: Well, this also mean adding it to `ConfigJson` reading and writing methods, and editing unit tests. I can do it, but for something useless, is it worth it?

Comment (virtualtam) [inline tests/utils/config/configJson.json.php]: Users are expected to read and edit these configuration files, so it could be worth it. I'm OK with keeping the matter for a future issue/PR, though :)

Comment (ArthurHoaro) [inline tests/utils/config/configJson.json.php]: Makes sense. I'll do it in this PR.

**Comment by virtualtam** [comment]
OK for me, ping @nodiscc ;-)

**Comment by ArthurHoaro** [comment]
Let's go!

**Comment by kalvn** [comment]
Nice work!

I noticed a typo in `configure.html` template where `$private_links_default` is used instead of what I believe will be `$disable_session_protection` on line 39.

I keep migrating my theme and will let you know if I find something else :)

**Comment by ArthurHoaro** [comment]
Thanks for the feedback, I'll fix that!
